Support using a fixed port number. Removed (broken) support for client and added s…

StinkySteve opened the pull request:
I ran into some issues with getting Webmock/VCR to behave nicely with this gem. By pinning the port we can use VCR to record the responses. This speeds up tests(and makes it less of a hassle to configure VCR. 

I've also broken the logic up into objects and added some tests.

frenkel commented:
Sorry, but this code will fail when I have 4 unicorn workers that al try to start Chrome on the same port at the same time. The random port should be kept to support this.

StinkySteve commented:
> Sorry, but this code will fail when I have 4 unicorn workers that al try to start Chrome on the same port at the same time. The random port should be kept to support this.

Random ports are still the default behaviour.

> What part of the client support is broken?

This part: 
```
     if options[:client].nil?
      # random port at 1025 or higher
      random_port = 1024 + Random.rand(65_535 - 1024)
      pid = Process.spawn "chrome --remote-debugging-port=#{random_port} --headless"
    end
    
    success = false
    10.times do
      begin
        TCPSocket.new('localhost', random_port)
        success = true 
        break
      rescue
      end
      sleep 1
    end
    
    return unless success
```

>  I think you should be able to use that in tests if you really want to mock at the network level. I would advise you to use mocha on a higher level though.

I'm testing it on a higher level. I'm using an adapter pattern to be able to user recorded VCR responses in development. Its really nice. I could elaborate on it if you are interested.

StinkySteve commented:
@frenkel I've added even more changes. I destroyed quite a bit of code during the refactor. I think everyting runs now and added a readme for future me. 
If you want to run the gem in isolation you can run the script in bin/run_with_website. I haven't tried this in your usecase yet. Let me know if I missed something!

StinkySteve commented:
I'm not sure whats going on but it looks like chrome is spawning some helpers that won't die when I kill them. I think this branch still needs more debugging.

StinkySteve commented:
@frenkel I think this PR is done. Let me know if I missed something!

frenkel commented:
I'll have a look tomorrow.

frenkel commented on lib/postdoc/client.rb:
```diff
@@ -0,0 +1,44 @@
+        header_template: false,
+        footer_template: false,
+        **options)
+      client.send_cmd 'Page.enable'
```
I get: `undefined method `send_cmd' for nil:NilClass ` on this line in multiple projects when switching from the latest release to this branch. These use `render pdf: 'show'` in the controller, in the same way as `test/dummy` does.

StinkySteve replied:
I'll look into it!

StinkySteve commented:
Sorry for the delay. I've solved the issue for the dummy app. Does it work in the your host apps as well?

frenkel commented:
Seems to work fine now, thanks!

frenkel commented:
After some more testing I discovered that it will print the default browser header/footer text such as the url and page number. Any idea what causes that? I've yanked the release for now. Do you not get those? It seems the logic for passing `displayHeaderFooter` is now different/wrong. In my use case I'm not passing any of those, so they will probably be nil.

StinkySteve commented:
Sorry for that. I think I noticed on the way and has a similar issue with the way it was implemented in the host application. I'll reproduce and solve it asap!
